freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaCloudStorageValidationService.java

```diff
@@ -0,0 +1,80 @@
+import com.sequenceiq.freeipa.service.GatewayConfigService;
+import com.sequenceiq.freeipa.service.freeipa.config.FreeIpaSaltConfigService;
+import com.sequenceiq.freeipa.service.stack.StackService;
+
+@Service
+public class FreeIpaCloudStorageValidationService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaCloudStorageValidationService.class);
+
+    @Inject
+    private HostOrchestrator hostOrchestrator;
+
+    @Inject
+    private GatewayConfigService gatewayConfigService;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private FreeIpaSaltConfigService freeIpaSaltConfigService;
+
+    public void validate(Long stackId) throws CloudbreakOrchestratorException {
+        Stack stack = stackService.getByIdWithListsInTransaction(stackId);
+        Set<InstanceMetaData> instanceMetaDatas = stack.getNotDeletedInstanceMetaDataSet();
+        List<GatewayConfig> gatewayConfigs = gatewayConfigService.getGatewayConfigs(stack, instanceMetaDatas);
+        Set<Node> allNodes = mapInstancesToNodes(instanceMetaDatas);
+
+        createAndPushPillarsToNodes(stackId, stack, gatewayConfigs, allNodes);
+        validateCloudStorage(stackId, stack, gatewayConfigs, allNodes);
+    }
+
+    private void validateCloudStorage(Long stackId, Stack stack, List<GatewayConfig> gatewayConfigs, Set<Node> allNodes) throws CloudbreakOrchestratorException {
+        GatewayConfig primaryGatewayConfig = gatewayConfigService.getPrimaryGatewayConfig(stack);
+        hostOrchestrator.validateCloudStorage(primaryGatewayConfig, gatewayConfigs, allNodes, new StackBasedExitCriteriaModel(stackId));
+    }
+
+    private void createAndPushPillarsToNodes(Long stackId, Stack stack, List<GatewayConfig> gatewayConfigs, Set<Node> allNodes)
+            throws CloudbreakOrchestratorFailedException {
+        SaltConfig saltConfig = freeIpaSaltConfigService.getSaltConfig(stack, allNodes);
+        hostOrchestrator.initSaltConfig(gatewayConfigs, allNodes, saltConfig, new StackBasedExitCriteriaModel(stackId));
+    }
+
+    private Set<Node> mapInstancesToNodes(Set<InstanceMetaData> instanceMetaDatas) throws CloudbreakOrchestratorFailedException {
+        Set<Node> allNodes = instanceMetaDatas.stream()
+                .map(im -> new Node(im.getPrivateIp(), im.getPublicIp(), im.getInstanceId(),
+                        im.getInstanceGroup().getTemplate().getInstanceType(), im.getDiscoveryFQDN(), "testGroup"))
+                .collect(Collectors.toSet());
+        validateNodeIsPresent(allNodes);
+        return allNodes;
+    }
+
+    private void validateNodeIsPresent(Set<Node> allNodes) throws CloudbreakOrchestratorFailedException {
+        if (allNodes.isEmpty()) {
+            String errorMessage = "There are no nodes to validate FreeIPA cloud storage.";
+            LOGGER.error(errorMessage);
+            throw new CloudbreakOrchestratorFailedException(errorMessage);
+        }
+    }
```

Comment:
this part looks like duplication from `FreeIpaInstallService`

Reply:
Good point. I cleaned up a few other places that had similar logic.